packages/core/src/document.ts

```diff
@@ -580,39 +581,73 @@ class Document extends EventEmitter implements DocStateHolder {
     await this.dispatcher.publishTip(this.id, this.tip)
   }
 
+  /**
+   * Returns whether the given CID is in the log, and there is a later log entry that is an anchor commit
+   * @param cid
+   */
+  async _isCidAnchored(cid: CID): Promise<boolean> {
+    let foundCid = false
+    for (const entry of this.state.log) {
+      if (entry.cid.equals(cid)) {
+        foundCid = true
+      }
+
+      if (foundCid && entry.type == CommitType.ANCHOR) {
+        return true
+      }
+    }
+    return false
+  }
+
   /**
    * Request anchor for the latest document state
    */
   async anchor (): Promise<void> {
-    this._context.anchorService.on(this.id.toString(), async (asr: AnchorServiceResponse): Promise<void> => {
+    // eslint-disable-next-line @typescript-eslint/no-this-alias
+    const doc = this
+    const tip: CID = this.tip
+
+    this._context.anchorService.on(this.id.toString(), async function listener (asr: AnchorServiceResponse) {
+      if (!asr.cid.equals(tip)) {
+        // This message is about a different tip for the same document
+        return
+      }
       switch (asr.status) {
         case 'PENDING': {
-          const state = this._doctype.state
+          const state = doc._doctype.state
           state.anchorScheduledFor = asr.anchorScheduledFor
-          this._doctype.state = state
-          await this._updateStateIfPinned()
+          doc._doctype.state = state
+          await doc._updateStateIfPinned()
           return
         }
         case 'PROCESSING': {
-          const state = this._doctype.state
+          const state = doc._doctype.state
           state.anchorStatus = AnchorStatus.PROCESSING
-          this._doctype.state = state
-          await this._updateStateIfPinned()
+          doc._doctype.state = state
+          await doc._updateStateIfPinned()
           return
         }
         case 'COMPLETED': {
-          await this._handleTip(asr.anchorRecord)
-          await this._updateStateIfPinned()
-          await this._publishTip()
+          doc._context.anchorService.removeListener(doc.id.toString(), listener)
 
-          this._context.anchorService.removeAllListeners(this.id.toString())
+          await doc._handleTip(asr.anchorRecord)
+          await doc._updateStateIfPinned()
+          await doc._publishTip()
           return
         }
         case 'FAILED': {
-          const state = this._doctype.state
+          doc._context.anchorService.removeListener(doc.id.toString(), listener)
+
+          if (await doc._isCidAnchored(tip)) {
+            // Even though the anchor request for this specific CID came back as FAILED, the cid
+            // still appears to be anchored.  This means that a later CID built on top of this
+            // one was anchored instead. In that case we should do nothing and simply return since
+            // the later CID that was successfully anchored should have already updated the state appropriately
+            return
+          }
```

Comment:
This is the condition I was referring to. Seems like if we check that `asr.cid` equals `tip` then this will never happen?

Reply:
This could happen if the notifications from the anchor service come in out of order. Basically if you learn about the later CID being anchored, and *then* get notified about the earlier CID failing.  You don't want to transition to FAILED in that case

Reply:
makes sense, thanks